### DISCUSSION TITLE
🐞 Elixir/Rust/Python: strange experimentalWithAllGPUs function name

### DISCUSSION THREAD

**Discussion by wingyplus** [dagger/dagger#12403] [Help]
### What is the issue?

See https://hexdocs.pm/dagger/Dagger.Container.html#experimental_with_all_gp_us/1. The function name should be `experimental_with_all_gpus`, `experimental_with_all_gp_us`

### Dagger version

dagger v0.9.4

**Comment by jedevc** [comment]
cc @vito since you've been in the graphql schema space :smile: 

I think this is some weirdness with how the string conversion library we use works, `GPUs` gets split into `GP`+`Us` instead of keeping it as is.

**Comment by vito** [comment]
My guess is we need to use [`strcase.ConfigureAcronym`](https://pkg.go.dev/github.com/iancoleman/strcase#ConfigureAcronym)

**Comment by wingyplus** [comment]
I workaround it by replacing it with `Gpus` (https://github.com/wingyplus/dagger/blob/elixir-new-codegen/sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex#L21). 😂

I found this issue on the Rust SDK  also https://github.com/dagger/dagger/blob/9bf9f7f303b6768efbba540376a0bd6eae6979a3/sdk/rust/crates/dagger-sdk/src/gen.rs#L889.

**Comment by jedevc** [comment]
Eek we also have this issue in python (cc @helderco): https://github.com/dagger/dagger/blob/2999573c5a34e85b8baac8c0150881d9c08a86b8/sdk/python/src/dagger/client/gen.py#L501

Poked into this - it looks like this is left up to the SDK codegen to convert, the input at the point of generation is `experimentalWithAllGPUs`.

In the rust case (which I'm more familiar with), this function performs the incorrect conversion: https://github.com/dagger/dagger/blob/2999573c5a34e85b8baac8c0150881d9c08a86b8/sdk/rust/crates/dagger-codegen/src/rust/functions.rs#L17-L19

Super weird though, because, if I change the code slightly:

```rust
pub fn format_struct_name(s: &str) -> String {
    let conv = Converter::new()
        .remove_boundary(Boundary::Acronym)
        .to_case(Case::Snake);
    conv.convert(s)
}
```

Then I get:

```diff
diff --git a/sdk/rust/crates/dagger-sdk/src/gen.rs b/sdk/rust/crates/dagger-sdk/src/gen.rs
index 7aa880b6e..5f9803f3c 100644
--- a/sdk/rust/crates/dagger-sdk/src/gen.rs
+++ b/sdk/rust/crates/dagger-sdk/src/gen.rs
@@ -929,7 +929,7 @@ impl Container {
     /// EXPERIMENTAL API! Subject to change/removal at any time.
     /// Configures all available GPUs on the host to be accessible to this container.
     /// This currently works for Nvidia devices only.
-    pub fn experimental_with_all_gp_us(&self) -> Container {
+    pub fn experimental_with_all_gpus(&self) -> Container {
         let query = self.selection.select("experimentalWithAllGPUs");
         return Container {
             proc: self.proc.clone(),
@@ -3195,7 +3195,7 @@ impl GeneratedCode {
         query.execute(self.graphql_client.clone()).await
     }
     /// Set the list of paths to mark generated in version control.
-    pub fn with_vcs_generated_paths(&self, paths: Vec<impl Into<String>>) -> GeneratedCode {
+    pub fn with_vcsgenerated_paths(&self, paths: Vec<impl Into<String>>) -> GeneratedCode {
         let mut query = self.selection.select("withVCSGeneratedPaths");
         query = query.arg(
             "paths",
@@ -3208,7 +3208,7 @@ impl GeneratedCode {
         };
     }
     /// Set the list of paths to ignore in version control.
-    pub fn with_vcs_ignored_paths(&self, paths: Vec<impl Into<String>>) -> GeneratedCode {
+    pub fn with_vcsignored_paths(&self, paths: Vec<impl Into<String>>) -> GeneratedCode {
         let mut query = self.selection.select("withVCSIgnoredPaths");
         query = query.arg(
             "paths",
```

It seems like when we transform to camel case, you can't have both:
- "VCSIgnore" -> "VCS" "Ignore"
- "GPUs" -> "GPUs"

**Comment by vito** [comment]
Welp. Maybe it's time to just stop using acronyms?

* WithVCSIgnoredPaths => WithIgnoredCodePaths
* WithVCSGeneratedPaths => WithGeneratedCodePaths
* ExperimentalWithAllGPUs => ExperimentalWithAllGraphicsDevices

We could alternatively maintain a bespoke list of acronyms and hope each language has a library that supports that sort of thing, but keeping that in sync across SDKs seems like it'd be a constant battle.

**Comment by wingyplus** [comment]
> Welp. Maybe it's time to just stop using acronyms?
> 
> * WithVCSIgnoredPaths => WithIgnoredCodePaths
> * WithVCSGeneratedPaths => WithGeneratedCodePaths
> * ExperimentalWithAllGPUs => ExperimentalWithAllGraphicsDevices
> 
> We could alternatively maintain a bespoke list of acronyms and hope each language has a library that supports that sort of thing, but keeping that in sync across SDKs seems like it'd be a constant battle.

Or we could use acronym word with a camel case, such as `experimentalWithAllGPUs` -> `experimentalWithAllGpus`. I’m not sure if it make sense to some language such as Go or Java that uses camel case style but it solve on snake case function like this issue.

**Comment by jedevc** [comment]
> We could alternatively maintain a bespoke list of acronyms

Even more alternatively, could we potentially add a method to the dagger API that performs the word splitting? Then each SDK joins them up? That way, we only need to handle this complexity in the Go SDK.

Though the issue with that is that the SDKs are using the graphql introspection API - so not 100% sure about that, if we were instead using the fancy new TypeDefs API, it would be super easy to add into that.

Just another option to add to the pile :tada:

**Comment by helderco** [comment]
Yeah, that's my plan with:
- https://github.com/dagger/dagger/issues/5226

Need to update the description, but have been thinking on modularizing it.
